### PR TITLE
Validate usages of assert*-methods in TypeInferenceTestCase

### PR DESCRIPTION
**Pull request by staabm**
as discussed in https://github.com/phpstan/phpstan-src/pull/2107/files#r1048450984

tested using

```diff
diff --git a/tests/PHPStan/Analyser/data/implode.php b/tests/PHPStan/Analyser/data/implode.php
index b2e2b9aa7..86a03a838 100644
--- a/tests/PHPStan/Analyser/data/implode.php
+++ b/tests/PHPStan/Analyser/data/implode.php
@@ -2,7 +2,8 @@

 namespace ImplodeFunctionReturn;

-use function PHPStan\Testing\assertType;
+use function PHPStan\Testing\assertNativeType;
+use function PHPStan\wrong\assertType;

 class Foo
 {
@@ -10,6 +11,8 @@ class Foo
        const ONE = 1;

        public function constants() {
+               assertNativeType('string', implode(',', [self::X, '345']));
+
                assertType("'12345'", implode(['12', '345']));

                assertType("'12345'", implode('', ['12', '345']));
```

```
➜  phpstan-src git:(pr/2118) ✗ vendor/bin/phpunit tests/PHPStan/Analyser/NodeScopeResolverTest.php
Xdebug: [Step Debug] Could not connect to debugging client. Tried: 127.0.0.1:9003 (through xdebug.client_host/xdebug.client_port) :-(
PHPUnit 9.5.23 #StandWithUkraine

Warning:       XDEBUG_MODE=coverage or xdebug.mode=coverage has to be set

E                                                                   1 / 1 (100%)

Time: 00:00.017, Memory: 54.00 MB

There was 1 error:

1) Error
The data provider specified for PHPStan\Analyser\NodeScopeResolverTest::testFileAsserts is invalid.
PHPUnit\Framework\AssertionFailedError: ERROR: Assert-Method PHPStan\wrong\assertType imported with wrong namespace called from line 16.
/Users/staabm/workspace/phpstan-src/src/Testing/TypeInferenceTestCase.php:182
/Users/staabm/workspace/phpstan-src/src/Node/ClassStatementsGatherer.php:115
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:553
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:3037
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:1751
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:618
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:294
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:569
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:294
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:668
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:294
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:638
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:249
/Users/staabm/workspace/phpstan-src/src/Testing/TypeInferenceTestCase.php:71
/Users/staabm/workspace/phpstan-src/src/Testing/TypeInferenceTestCase.php:197
/Users/staabm/workspace/phpstan-src/tests/PHPStan/Analyser/NodeScopeResolverTest.php:19

ERRORS!
Tests: 1, Assertions: 0, Errors: 1.
```

----

```diff
diff --git a/tests/PHPStan/Analyser/data/implode.php b/tests/PHPStan/Analyser/data/implode.php
index b2e2b9aa7..8e80ecead 100644
--- a/tests/PHPStan/Analyser/data/implode.php
+++ b/tests/PHPStan/Analyser/data/implode.php
@@ -2,7 +2,7 @@

 namespace ImplodeFunctionReturn;

-use function PHPStan\Testing\assertType;
+use function PHPStan\Testing\assertNativeType;

 class Foo
 {
@@ -10,6 +10,8 @@ class Foo
        const ONE = 1;

        public function constants() {
+               assertNativeType('string', implode(',', [self::X, '345']));
+
                assertType("'12345'", implode(['12', '345']));

                assertType("'12345'", implode('', ['12', '345']));
```

```
➜  phpstan-src git:(asserts) ✗ vendor/bin/phpunit tests/PHPStan/Analyser/NodeScopeResolverTest.php
PHPUnit 9.5.23 #StandWithUkraine

Warning:       No code coverage driver available

E                                                                   1 / 1 (100%)

Time: 00:00.011, Memory: 52.00 MB

There was 1 error:

1) Error
The data provider specified for PHPStan\Analyser\NodeScopeResolverTest::testFileAsserts is invalid.
PHPUnit\Framework\AssertionFailedError: ERROR: Missing import for assertType() on line 15.
/Users/staabm/workspace/phpstan-src/src/Testing/TypeInferenceTestCase.php:131
/Users/staabm/workspace/phpstan-src/src/Node/ClassStatementsGatherer.php:115
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:553
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:3037
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:1751
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:618
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:294
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:569
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:294
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:668
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:294
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:638
/Users/staabm/workspace/phpstan-src/src/Analyser/NodeScopeResolver.php:249
/Users/staabm/workspace/phpstan-src/src/Testing/TypeInferenceTestCase.php:71
/Users/staabm/workspace/phpstan-src/src/Testing/TypeInferenceTestCase.php:197
/Users/staabm/workspace/phpstan-src/tests/PHPStan/Analyser/NodeScopeResolverTest.php:19

ERRORS!
Tests: 1, Assertions: 0, Errors: 1.
```

### CODE REVIEW

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
This needs a test. Please add a new class that extend TypeInferenceTestCase and do something with `$this->expectException('PHPUnit\Framework\ExpectationFailedException');` there (https://stackoverflow.com/questions/3917909/how-to-indicate-that-a-phpunit-test-is-expected-to-fail). Thanks.

**Comment by staabm** [comment]
@herndlm looks like the new patching solution has problems on windows with php 7.2

File: tests/PHPStan/Tests/TypeInferenceTestCaseTest.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Tests;
```
Comment (ondrejmirtes) [inline tests/PHPStan/Tests/TypeInferenceTestCaseTest.php]: This namespace could be `Testing`.

File: tests/PHPStan/Tests/TypeInferenceTestCaseTest.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		$this->expectException(AssertionFailedError::class);
+		$this->expectExceptionMessage('ERROR: Assert-Method SomeWrong\Namespace\assertNativeType imported with wrong namespace called from line 8.');
+
+		$this->gatherAssertTypes(__DIR__ . '/data/assert-native-type-wrong-namespace.php');
+	}
+
+	public function testWrongAssertTypeNamespace(): void
+	{
+		$this->expectException(AssertionFailedError::class);
+		$this->expectExceptionMessage('ERROR: Assert-Method SomeWrong\Namespace\assertType imported with wrong namespace called from line 8.');
```
Comment (ondrejmirtes) [inline tests/PHPStan/Tests/TypeInferenceTestCaseTest.php]: The error messages could be cleaned up a bit.

1) `ERROR: ` at the beginning isn't needed, it's obvious when the test fails that it's an error.
2) "on line" 8, not "from line"
3) The message should read: "Function PHPStan\Testing\assertType imported with wrong namespace SomeWrong\Namespace..."

File: src/Testing/TypeInferenceTestCase.php [hunk 3]
```diff
@@ -125,7 +127,13 @@ public static function gatherAssertTypes(string $file): array
 			}
 
 			$functionName = $nameNode->toString();
-			if ($functionName === 'PHPStan\\Testing\\assertType') {
+			if (in_array($functionName, ['assertType', 'assertNativeType', 'assertVariableCertainty'], true)) {
```
Comment (ondrejmirtes) [inline src/Testing/TypeInferenceTestCase.php]: This should be case-insensitive + please also test it

File: src/Testing/TypeInferenceTestCase.php [hunk 4]
```diff
@@ -125,7 +127,13 @@ public static function gatherAssertTypes(string $file): array
...
+				self::fail(sprintf(
+					'ERROR: Missing import for %s() on line %d.',
```
Comment (ondrejmirtes) [inline src/Testing/TypeInferenceTestCase.php]: It's not import but "use statement".

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: tests/PHPStan/Tests/TypeInferenceTestCaseTest.php (hunk 1 above)
Comment (staabm) [reply tests/PHPStan/Tests/TypeInferenceTestCaseTest.php]: already existing test-classes in the same folder use `PHPStan\Tests`.

`Testing` is used in the `src/` folder

Comment (ondrejmirtes) [reply tests/PHPStan/Tests/TypeInferenceTestCaseTest.php]: But here you're testing **PHPStan\Testing\TypeInferenceTestCase** class.

**Comment by ondrejmirtes** [comment]
Fixed the Windows problem https://github.com/phpstan/phpstan-src/commit/2df5f879aa37fee7afb23f92714832dddc30191d

File: src/Testing/TypeInferenceTestCase.php (hunk 3 above)
Comment (staabm) [reply src/Testing/TypeInferenceTestCase.php]: done

File: tests/PHPStan/Tests/TypeInferenceTestCaseTest.php (hunk 2 above)
Comment (staabm) [reply tests/PHPStan/Tests/TypeInferenceTestCaseTest.php]: 1. nearly all other error messages in TypeInferenceTestCase use the `ERROR: ` prefix - stripped it
2. done
3. done

**Comment by staabm** [comment]
Looking at it again, I think a dataProvider could be useful

**Comment by ondrejmirtes** [comment]
Yes :)

**Comment by ondrejmirtes** [comment]
Thank you.

**Pull request merged**